// @allurereport/plugin-awesome/dist/generateTimeline.js
import { hasLabels } from "@allurereport/web-commons";
const DEFAULT_MIN_DURATION = 1;
const HOST_LABEL = "host";
const THREAD_LABEL = "thread";
const DEFAULT_TIMELINE_OPTIONS = {
    minDuration: DEFAULT_MIN_DURATION,
};
export const generateTimeline = async (writer, store, options) => {
    const { timeline = DEFAULT_TIMELINE_OPTIONS } = options;
    const { minDuration = DEFAULT_MIN_DURATION } = timeline;
    const testResults = await store.allTestResults({ includeHidden: true });
    const result = [];
    for (const test of testResults) {
        const hasStart = Number.isInteger(test.start);
        const hasStop = Number.isInteger(test.stop);
        if (!hasStart || !hasStop) {
            continue;
        }
        const duration = test.duration ?? test.stop - test.start;
        if (duration < minDuration) {
            continue;
        }
        if (!hasLabels(test, [HOST_LABEL, THREAD_LABEL])) {
            continue;
        }
        result.push({
            id: test.id,
            historyId: test.historyId,
            name: test.name,
            status: test.status,
            hidden: test.hidden,
            labels: test.labels,
            environment: test.environment,
            start: test.start,
            stop: test.stop,
            duration,
        });
    }
    await writer.writeWidget("timeline.json", result);
};
