// @allurereport/web-commons/dist/charts/generateStabilityDistributionChart.js
import { ChartType } from "@allurereport/charts-api";
import { createHashStorage, createMapWithDefault } from "./utils.js";
const DEFAULT_THRESHOLD = 90;
const DEFAULT_GROUP_BY = "feature";
const CUSTOM_LABEL_NAME_PREFIX = "label-name:";
const getTrLabelValue = (testResult, labelName) => {
    return testResult.labels.find((label) => label.name === labelName)?.value;
};
const getStabilityRate = (passed, total) => {
    return Math.floor((passed / total) * 10000) / 100;
};
const NON_SIGNIFICANT_STATUSES = ["unknown", "skipped"];
class IncludeAllArray extends Array {
    constructor() {
        super();
    }
    includes() {
        return true;
    }
}
const allValuesList = new IncludeAllArray();
export const generateStabilityDistributionChart = (props) => {
    const { options, storeData } = props;
    const { title, threshold = DEFAULT_THRESHOLD, skipStatuses = NON_SIGNIFICANT_STATUSES, groupBy = DEFAULT_GROUP_BY, groupValues = [], } = options;
    const { testResults } = storeData;
    const labelName = groupBy.startsWith(CUSTOM_LABEL_NAME_PREFIX)
        ? groupBy.slice(CUSTOM_LABEL_NAME_PREFIX.length)
        : groupBy;
    const labelValuesList = groupValues.length > 0 ? groupValues : allValuesList;
    const trsStatsByLabelValues = createMapWithDefault({
        passed: 0,
        total: 0,
    });
    const keys = {};
    const hashes = createHashStorage();
    for (const tr of testResults) {
        const labelValue = getTrLabelValue(tr, labelName);
        if (!labelValue) {
            continue;
        }
        if (!labelValuesList.includes(labelValue)) {
            continue;
        }
        const labelValueHash = hashes.get(labelValue);
        keys[labelValueHash] = labelValue;
        if (skipStatuses.includes(tr.status)) {
            continue;
        }
        trsStatsByLabelValues.get(labelValueHash).total++;
        if (tr.status === "passed") {
            trsStatsByLabelValues.get(labelValueHash).passed++;
        }
    }
    return {
        data: trsStatsByLabelValues.entries.map(([id, { passed, total }]) => ({
            id,
            stabilityRate: total > 0 ? getStabilityRate(passed, total) : 0,
        })),
        keys,
        type: ChartType.StabilityDistribution,
        title,
        threshold,
    };
};
